pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -140,6 +142,8 @@ public class PinotLLCRealtimeSegmentManager {
   // Max time to wait for all LLC segments to complete committing their metadata while stopping the controller.
   private static final long MAX_LLC_SEGMENT_METADATA_COMMIT_TIME_MILLIS = 30_000L;
 
+  private Map<Pair<String, String>, SegmentErrorInfo> _errorCache;
```

Comment:
This is not really a cache because we will never clean up the entries as of now. Do you plan to add the cleanup logic in this PR? If not, let's add a TODO at least

Reply:
Im not sure if this is meant to be cleaned up, since we need keep it around for the debug api calls ?

Reply:
I should limit its size. I am thinking of using a `LinkedHashMap` to limit its entry. Similar to https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html#removeEldestEntry-java.util.Map.Entry-

Is that OK? 

However, these methods are not synchronized. I can use [Collections.synchronizedMap](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedMap-java.util.Map-) to wrap the LinkedHashMap. 

Also another minor improvement is to change the signature to `getSegmentErrors(String tableName, String segmentName)`. Then iterators and streams are not required and all the access will be through `get` which is synchronized.

Reply:
Just double checking my understanding of the error cache.  It's a map from each (table, segment) pair that is on this server to the most recent error message that was seen for that table/segment? In other words, for each server, we'll see the most recent error on each segment on that server. 

1. Longer term is how to manage noisy errors vs not-noisy errors.  For example: if there's an error with missing offsets (which you're monitoring for in this PR) and a decoding error on 1/5 messages, the decoding error will flood the cache and block out the Offset Error from being seen.
2. What happens when a table/segment is deleted or moved? The error cache will still have the non-existent segments and provide invalid information.  We have this issue with Ingestion Lag metrics and it's frequently causing false alerts and issues. If this happens multiple times then we can wind up with many servers reporting errors for the same segment which will be confusing during investigations.
3. If you limit the size of this map, then it still needs to support all the extant segments that are on a server: so I'm not sure setting a fixed limit will work b/c how many segments a single server can have is not, so far as I know, strictly limited. So how can we determine what the max size should be?

Reply:
One point is that this is for debug APIs only. Errors from this list is not exposed to the user. The official user interface is metrics and logs. Based on the fact that the error list is for debugging only:


1.  High error rate: Limiting the size is a good idea so that it doesnt use up memory. Losing errors is OK as this is not the source of truth and only for debugging. Right now there is only one error source. If there are too many, the important aspect is that a data loss occurred.
2. Segment lifecycle: Since it is not the source of truth and used by devs only, mismatch is OK. The main contribution of the PR is the metric to track data loss and the log with all the necessary info.